### PR TITLE
Google contacts and calendar

### LINKED ISSUE #8: Google OAuth for DAV
https://developers.google.com/google-apps/calendar/auth

Apparently we have to register the client at Google, not sure how we would keep the keys secret (if any) in an opensource project.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1489356-google-oauth-for-dav?utm_campaign=plugin&utm_content=tracker%2F574345&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F574345&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by untitaker**
- [x] Proper documentation
- [x] With the refactor, `get_storage_init_args` now misses a lot of arguments for all DAV and Google storage types

Fix #8

### CODE REVIEW

**Comment by untitaker** [comment]
@marmarek I've added you as a collaborator!

So, I currently get this error after pasting the `refresh_token` back into the config:

	  File "/home/untitaker/venvs/vdirsyncer/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 376, 
	in parse_token_response
		validate_token_parameters(params)
	  File "/home/untitaker/venvs/vdirsyncer/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 383, 
	in validate_token_parameters
		raise_from_error(params.get('error'), params)
	  File "/home/untitaker/venvs/vdirsyncer/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 271, in r
	aise_from_error
		raise cls(**kwargs)
	oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) Token has been revoked.
	critical: Failed to initialize cal
	critical:   - (invalid_grant) Token has been revoked.

File: vdirsyncer/storage/google.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from .. import exceptions, utils
+
+logger = logging.getLogger(__name__)
+
+
+TOKEN_URL = 'https://accounts.google.com/o/oauth2/v2/auth'
+REFRESH_URL = 'https://www.googleapis.com/oauth2/v4/token'
+
+CLIENT_ID = '155040592229-mth5eiq7qt9dtk46j0vcnoometuab9mb.apps.googleusercontent.com'
+CLIENT_SECRET = 'flVXF7jB2A2-YC4rg2sUCCX1'
```
Comment (untitaker) [inline vdirsyncer/storage/google.py]: gcalcli does the same thing. I wonder if we get banned for this.

File: vdirsyncer/storage/google.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        self.useragent = client_id
+        self._settings = {}
+
+        if not have_oauth2:
+            raise exceptions.UserError('requests-oauthlib not installed')
+
+        if refresh_token:
+            token = {
+                'refresh_token': refresh_token,
+                'access_token': 'dummy',
```
Comment (untitaker) [inline vdirsyncer/storage/google.py]: So I'm unsure whether this part works as it should, and I'm now unsure whether just requesting a new access token was a good idea after all.

**Comment by untitaker** [comment]
Ok, calendar stuff is working now, as far as I can see. That means, full read-write access to calendars. @marmarek I've reverted to using a file, sorry for annoyance :(

However, contacts don't work for me at all, everything returns a 404.

**Comment by marmarek** [comment]
I've got similar problem when specified only base URL. It works with full contacts URL (in form `https://www.googleapis.com/carddav/v1/principals/EMAIL/lists/default` as I've written in docs). Unfortunately haven't managed managed to get it working with automatic discovery.
According to [google documentation](https://developers.google.com/google-apps/carddav/) it may work with `https://www.googleapis.com/.well-known/carddav`, but using PROPFIND request, not GET as it is currently done.

**Comment by untitaker** [comment]
@homu r+

**Comment by geier** [comment]
I'm pretty excited that vdirsyncer now supports Google's OAuth! While I probably won't use it myself, this has been one of the most requested features (for vdirsyncer and khal). Thanks @marmarek!

**Comment by untitaker** [comment]
Agreed, thanks from me too @marmarek!

**Comment by terlar** [comment]
Thanks for the great work, I have been looking forward to use the vdirsyncer, but at my company we use google calendar for our calendars.

I just tried this out and cannot get it to work, it complains about:

`critical: google_calendar storage doesn't take the parameters: instance_name`

However I have not set this parameter and setting it to `null` doesn't work either. It seems the code automatically adds this one, but then it is not allowed. Any idea what I could be doing wrong or if this is an actual issue?

Once again thanks for your effort :bow:

**Comment by untitaker** [comment]
Could you post the full config and output? Could be that the error message is just warped.

On 5 April 2016 14:39:38 CEST, Terje Larsen <notifications@github.com> wrote:
>Thanks for the great work, I have been looking forward to use the
>vdirsyncer, but at my company we use google calendar for our calendars.
>
>I just tried this out and cannot get it to work, it complains about:
>
>`critical: google_calendar storage doesn't take the parameters:
>instance_name`
>
>However I have not set this parameter and setting it to `null` doesn't
>work either. It seems the code automatically adds this one, but then it
>is not allowed. Any idea what I could be doing wrong or if this is an
>actual issue?
>
>Once again thanks for your effort :bow: 
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/399#issuecomment-205781482

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by terlar** [comment]
config:
```ini
[general]
status_path = ~/.vdirsyncer/status/

[pair work_calendar]
conflict_resolution = b wins
a = work_calendar_local
b = work_calendar_remote
collections = ["from a", "from b"]
metadata = ["color", "displayname"]

[storage work_calendar_local]
type = filesystem
path = ~/.calendars/work/
fileext = .ics

[storage work_calendar_remote]
type = google_calendar
token_file = ~/.google-token-work
```

output:
```
$ vdirsyncer sync
Discovering collections for pair work_calendar
Opening [MASKED]
Follow the instructions on the page.
Paste obtained code: [MASKED]
critical: google_calendar storage doesn't take the parameters: instance_name

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/vdirsyncer/cli/utils.py", line 228, in _discover_from_config
    discovered = list(cls.discover(**config))
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 383, in discover
    session, _ = cls.session_class.init_and_remaining_args(**kwargs)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 301, in init_and_remaining_args
    return cls(**self_args), remainder
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 87, in __init__
    _save_token(token)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 49, in _save_token
    with open(token_file, 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: '~/.google-token-work'
critical: Failed to initialize work_calendar_remote
critical:   - [Errno 2] No such file or directory: '~/.google-token-work'
error: 1 out of 1 tasks failed.
```

I also see that it complains about the token file which is a bit weird, because it is there and I guess without it I wouldn't be able to get the obtained code from the website?

**Comment by terlar** [comment]
So seems it doesn't support the tilde `~`. If I put a full path it works, I guess I assumed it would work since it does for the status path and other configs. If I put an absolute path it worked.

**Comment by untitaker** [comment]
Thanks, I think I fixed both bugs in https://github.com/pimutils/vdirsyncer/pull/403. Could you test that PR? (since you were able to install from master, I guess other branches wouldn't be a problem?)

**Comment by untitaker** [comment]
I've now merged the PR, please redo any commands you used to install vdirsyncer from master.

**Comment by terlar** [comment]
Sorry, I didn't see your message until now. I have tried it out now again and I get some new errors.

output:
```
$ vdirsyncer sync
Syncing work_calendar/XXX
Syncing work_calendar/XXX
Syncing work_calendar/XXX
Syncing work_calendar/XXX
critical: google_calendar storage doesn't take the parameters: url

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/vdirsyncer/cli/utils.py", line 376, in storage_instance_from_config
    return cls(**new_config)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 128, in __init__
    **kwargs
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 701, in __init__
    super(CaldavStorage, self).__init__(**kwargs)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 375, in __init__
    super(DavStorage, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'url'
critical: google_calendar storage doesn't take the parameters: url
critical: Failed to initialize work_calendar_remote
critical:   - __init__() got an unexpected keyword argument 'url'
critical: google_calendar storage doesn't take the parameters: url

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/vdirsyncer/cli/utils.py", line 376, in storage_instance_from_config
    return cls(**new_config)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 128, in __init__
    **kwargs
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 701, in __init__
    super(CaldavStorage, self).__init__(**kwargs)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 375, in __init__
    super(DavStorage, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'url'
critical: google_calendar storage doesn't take the parameters: url

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/vdirsyncer/cli/utils.py", line 376, in storage_instance_from_config
    return cls(**new_config)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 128, in __init__
    **kwargs
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 701, in __init__
    super(CaldavStorage, self).__init__(**kwargs)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 375, in __init__
    super(DavStorage, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'url'
critical: Failed to initialize work_calendar_remote
critical:   - __init__() got an unexpected keyword argument 'url'

Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/vdirsyncer/cli/utils.py", line 376, in storage_instance_from_config
    return cls(**new_config)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/google.py", line 128, in __init__
    **kwargs
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 701, in __init__
    super(CaldavStorage, self).__init__(**kwargs)
  File "/usr/lib/python3.5/site-packages/vdirsyncer/storage/dav.py", line 375, in __init__
    super(DavStorage, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'url'
critical: Failed to initialize work_calendar_remote
critical:   - __init__() got an unexpected keyword argument 'url'
critical: Failed to initialize work_calendar_remote
critical:   - __init__() got an unexpected keyword argument 'url'
error: 4 out of 5 tasks failed.                                      
```

**Comment by untitaker** [comment]
Yikes, please try https://github.com/pimutils/vdirsyncer/pull/406

**Pull request merged**